src/reactionsystem.jl

```diff
@@ -0,0 +1,165 @@
+using Catalyst, ModelingToolkit, Symbolics
+
+# module SBML
+# struct Model end
+# struct Reaction end
+# struct Compartment end
+# struct Species end
+# end
+
+""" ReactionSystem constructor """
+function ModelingToolkit.ReactionSystem(model::Model; kwargs...)  # Todo: requires unique parameters (i.e. SBML must have been imported with localParameter promotion in libSBML)
+    model = make_extensive(model)
+    # model = expand_reversible(model)
+    rxs = mtk_reactions(model)
+    t = Catalyst.DEFAULT_IV
+    species = [create_var(k) for k in keys(model.species)]
+    params = vcat([create_param(k) for k in keys(model.parameters)], [create_param(k) for k in keys(model.compartments)])
+    ReactionSystem(rxs,t,species,params; kwargs...)
+end
+
+""" ReactionSystem constructor """
+function ModelingToolkit.ReactionSystem(sbmlfile::String; kwargs...)
+    model = readSBML(sbmlfile)
+    ReactionSystem(model; kwargs...)
+end
+
+""" ODESystem constructor """
+function ModelingToolkit.ODESystem(model::Model; kwargs...)
+    rs = ReactionSystem(model; kwargs...)
+    model = make_extensive(model)  # PL: consider making `make_extensive!` to avoid duplicate calling in ReactionSystem and here
+    u0map = get_u0(model)
+    parammap = get_paramap(model)
+    defaults = Dict(vcat(u0map, parammap))
+    convert(ODESystem, rs, defaults=defaults)
+end
+
+""" ODESystem constructor """
+function ModelingToolkit.ODESystem(sbmlfile::String; kwargs...)
```

Comment:
same for this dispatch, I think this is type-piracy https://docs.julialang.org/en/v1/manual/style-guide/#Avoid-type-piracy. It's not a big deal ATM though, so I'm fine with it

Reply:
Not a problem now, I think:
```
julia> using ModelingToolkit, Catalyst

julia> methods(ReactionSystem)
# 3 methods for type constructor:
[1] ReactionSystem(iv; kwargs...) in ModelingToolkit at C:\Users\wolf5212\.julia\packages\ModelingToolkit\Bb2DA\src\systems\reaction\reactionsystem.jl:161
[2] ReactionSystem(eqs, iv, species, params; observed, systems, name) in ModelingToolkit at C:\Users\wolf5212\.julia\packages\ModelingToolkit\Bb2DA\src\systems\reaction\reactionsystem.jl:152
[3] ReactionSystem(eqs, iv, states, ps, observed, name, systems) in ModelingToolkit at C:\Users\wolf5212\.julia\packages\ModelingToolkit\Bb2DA\src\systems\reaction\reactionsystem.jl:147

julia> methods(ODESystem)
# 6 methods for type constructor:
[1] ODESystem(eq::Equation, args...; kwargs...) in ModelingToolkit at C:\Users\wolf5212\.julia\packages\ModelingToolkit\Bb2DA\src\systems\diffeqs\odesystem.jl:226
[2] ODESystem(eqs::Vector{Equation}, iv::Sym, states::Vector{T} where T, ps::Vector{T} where T, observed::Vector{Equation}, tgrad::Base.RefValue{Vector{Num}}, jac::Base.RefValue{Any}, Wfact::Base.RefValue{Matrix{Num}}, 
Wfact_t::Base.RefValue{Matrix{Num}}, name::Symbol, systems::Vector{ODESystem}, defaults::Dict, structure, reduced_states::Vector{T} where T) in ModelingToolkit at C:\Users\wolf5212\.julia\packages\ModelingToolkit\Bb2DA\src\systems\diffeqs\odesystem.jl:26
[3] ODESystem(deqs::AbstractVector{var"#s6"} where var"#s6"<:Equation, iv, dvs, ps; observed, systems, name, default_u0, default_p, defaults) in ModelingToolkit at C:\Users\wolf5212\.julia\packages\ModelingToolkit\Bb2DA\src\systems\diffeqs\odesystem.jl:75
[4] ODESystem(eqs) in ModelingToolkit at C:\Users\wolf5212\.julia\packages\ModelingToolkit\Bb2DA\src\systems\diffeqs\odesystem.jl:137
[5] ODESystem(eqs, iv; kwargs...) in ModelingToolkit at C:\Users\wolf5212\.julia\packages\ModelingToolkit\Bb2DA\src\systems\diffeqs\odesystem.jl:137
[6] ODESystem(eqs, iv, states, ps, observed, tgrad, jac, Wfact, Wfact_t, name, systems, defaults, structure, reduced_states) in ModelingToolkit at C:\Users\wolf5212\.julia\packages\ModelingToolkit\Bb2DA\src\systems\diffeqs\odesystem.jl:26
```

Reply:
The real question is where reaction network importers like our SBML importer (but also CellML and BioNetGen importers) should go in the end. I am not yet convinced that having separate packages for every model specification language is the right way to go. What type of file is handled can be automatically inferred from the ending or header given a filename as String.
We could put the importers in:
- ModelingToolkit.jl: For instance, we could dispatch the filename as `String` to the `ReactionSystem` constructor: `ModelingToolkit.ReactionNetwork(filename::String)`
- ReactionNetworkImporters.jl: For instance, we could dispatch a `NetworkFileFormat` constructor such as `BNGNetwork()` and filename to `loadrxnnetwork`: `ReactionNetworkImporters.loadrxnetwork(BNGNetwork(), filename)`

@isaacsas: any thoughts?

Reply:
@shahriariravanian: what are your plans for CellMLToolkit. Do you think it should be moved under ReactionNetworkImporters.jl (see also Sam's and Chris' comment below)? Or shall it remain a standalone library?

Reply:
I think eventually it is a good idea to put all of them under the same package, but ReactionNetworkImporters doesn't seem like the right place for CellMLToolkit. 

First, CellMLToolkit mainly outputs ODESystems (and possibly DAESystems in the next version), not reaction networks. While CellMLToolkit specification can handle reaction networks, very few actual CellML files contain one (the same that SBML files can define an ODE, but most don't). For whatever historical reason, CellML and SBML have diverged to their own niches. 

Second, while it is currently only an importer, the future plan for CellMLToolkit is -- after adding DAEs and implementing units handling and Unitful integration -- to write an exporter. I don't know if you plan to add export functionality to SBML, but I feel that to use the full benefits of symbolic manipulation, we need to have the capability to output the results.

Reply:
@ChrisRackauckas @isaacsas @shahriariravanian @anandijain : After @shahriariravanian 's response that CellMLToolkit will grow bigger and does not fit well under the ReactionNetworkImporters umbrella I am inclined to have all Toolkits for biomodel importers as independent packages. Most users have their favourite model specification language and will ever only need one importer, so there are few reasons to combine them in one package. The only reason I can think of, is to make sure they have a similar API. But for the few users who use two or more model specification languages we can still take care that the API looks similar across importers.

To this end, I could imagine mimicking the constructors of the `MTK.AbstractSystem` types without overloading them to avoid type piracy. A bit like so (or is this a bad idea?):

```
module MTK
ODEsys(x) = println(x)
export ODEsys
end

module CellMLTK
using Main.MTK
ODEsys(x::String) = Main.MTK.ODEsys("Creating ODEsys from CellML file")
end

module SbmlTK
using Main.MTK
ODEsys(x::String) = Main.MTK.ODEsys("Creating ODEsys from SBML file")
end

using Main.MTK
x = "Creating ODEsys from scratch"
ODEsys(x)  # Creating ODEsys from scratch

x="Cannot create ODEsys from filename"
ODEsys(x)  # Cannot create ODEsys from filename
Main.CellMLTK.ODEsys(x)  # Creating ODEsys from CellML file
Main.SbmlTK.ODEsys(x)  # Creating ODEsys from SBML file
```

Reply:
Please don't shadow, just extend.

Reply:
To avoid misunderstandings, I assume shadowing is creating another function with the same name that takes precedence over the old function. And extending means overloading with a more specific type to make use of multiple dispatch. Is this correct?

Anyway, I do not get it to work. Especially not without avoiding type piracy.
```
module MTK
ODEsys(x) = println(x)
export ODEsys
end
module CellMLTK
using Main.MTK
Main.MTK.ODEsys(x::String) = ODEsys("Creating ODEsys from CellML file")
end
# module SbmlTK
# using Main.MTK
# Main.MTK.ODEsys(x::String) = ODEsys("Creating ODEsys from SBML file")
# export ODEsys
# end
using Main.MTK, Main.CellMLTK  #, Main.SbmlTK
x = "Creating ODEsys from scratch"
ODEsys(x)  # Creating ODEsys from scratch
```
```
ERROR: StackOverflowError:
Stacktrace:
 [1] ODEsys(x::String) (repeats 79984 times)
   @ Main.CellMLTK .\REPL[2]:3
```
Do you have any suggestions?